Feature/interactive app

musharraf-pg opened the pull request:
Added state / interactivity to app

musharraf-pg committed:
Implemented CRUD operations on todos

musharraf-pg committed:
Implemented batch update and delete actions

YouGoJoe commented on src/components/App.js:
```diff
@@ -31,37 +31,95 @@ function filterTodos(todos, filter) {
-      </ContentStyled>
-    </TodoAppStyled>
-  );
+const App = class extends React.Component {
```
typically this will just be `class App extends Component {...}`. Reads a little bit cleaner

YouGoJoe commented on src/components/App.js:
```diff
@@ -31,37 +31,95 @@ function filterTodos(todos, filter) {
-    </TodoAppStyled>
-  );
+const App = class extends React.Component {
+  constructor(props) {
```
if you're just setting state in the constructor, you can just move that out to the class definition like:
```
class myClass extends Component {
  state = {...some state here},
  ...
}
```
See [the parent app](https://github.com/SMARTeacher/prodigy-hydra/blob/develop/packages/app-parent/src/components/AppContainer.js#L40) for a quick example

YouGoJoe commented on src/components/App.js:
```diff
@@ -31,37 +31,95 @@ function filterTodos(todos, filter) {
+
+  onEnterNewTodo = (title) => {
+    this.setState({
+      todos: this.state.todos.concat([{ title: title, completed: false, editing: false, id: uniqueId() }])
```
This is a really common React gotcha. If you are using current state to inform future state, you'll want to [pass setState a function](https://reactjs.org/docs/react-component.html#setstate), not an object.[Here's](https://github.com/SMARTeacher/prodigy-react-toys/blob/develop/src/components/MiniCart.js#L102) an example from the toys app. 

You'll want to do something like:
```
this.setState(prevState => ({ todos: prevState.todos.concat...})
```

Also, is there a reason we are using `concat` instead of `push`?

musharraf-pg replied:
@YouGoJoe using concat since push would mutate prevState

YouGoJoe commented on src/components/App.js:
```diff
@@ -31,37 +31,95 @@ function filterTodos(todos, filter) {
+
+  onDeleteToDo = (todo) => {
+    this.setState({
+      todos: difference(this.state.todos, [todo])
```
same point about passing setState a funciton

YouGoJoe commented on src/components/App.js:
```diff
@@ -31,37 +31,95 @@ function filterTodos(todos, filter) {
+    });
+  };
+
+  // onToggleToDoComplete = (todo) => {
```
nit: delete commented out code

YouGoJoe commented on src/components/Header.js:
```diff
@@ -47,11 +47,36 @@ const NewTodoFieldStyled = styled.input`
+    constructor(props) {
+        super(props);
+
+        this.state = {
```
same point about state

YouGoJoe commented on src/components/Header.js:
```diff
@@ -47,11 +47,36 @@ const NewTodoFieldStyled = styled.input`
+    render() {
+        return (
+            <HeaderStyled>
+                <MarkAllCompleteStyled type="checkbox" checked={this.props.markAllCompleteChecked} onChange={() => this.props.onUpdateAllCompletedTo(!this.props.markAllCompleteChecked)}/>
```
nit: destructure at the top of `render()` like `const { markAllCompleteChecked, onUpdateAllCompletedTo, ...} = this.props;`

trevorwright replied:
Long line

YouGoJoe commented on src/components/ToDoItem.js:
```diff
@@ -53,27 +53,47 @@ const ToDoEditStyled = styled.input`
+    render() {
+        const {
+            props: { todo, onDeleteToDo, onToggleToDoComplete, onEditToDo, onStartEditTodo, onUpdateToDo },
+        } = this;
```
this can just be `... = this.props;` , saves you a nesting layer.

YouGoJoe commented on src/components/ToDoList.js:
```diff
@@ -3,6 +3,7 @@ import styled from 'styled-components';
 import ToDoItem from './ToDoItem';
 import PropTypes from 'prop-types';
 import ToDoType from '../types/ToDoType';
+import { omit } from 'lodash';
```
do we need this omit anymore?

YouGoJoe commented on src/components/App.js:
```diff
@@ -31,37 +31,95 @@ function filterTodos(todos, filter) {
+  // };
+
+  onUpdateToDo = (oldTodo, newTodo) => {
+    const newTodos = [ ...this.state.todos ];
```
same point about using a function with setState when you rely on the previous state. You may also be interested in an [immutable update pattern](https://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html#updating-an-item-in-an-array) for updating an item in an array, but it is by no means required for this sample app.

YouGoJoe reviewed (changes requested):
Looking really good! Just need to change how we're calling setState

trevorwright commented on src/components/App.js:
```diff
@@ -31,37 +31,95 @@ function filterTodos(todos, filter) {
+
+  onUpdateAllCompletedTo = (newCompletedStatus) => {
+    this.setState({
+      todos: this.state.todos.map(todo => ({...todo, completed: newCompletedStatus}))
```
This also should be using functional set state as Joe has previously outlined

trevorwright commented on src/components/App.js:
```diff
@@ -31,37 +31,95 @@ function filterTodos(todos, filter) {
+
+  onClearCompleted = () => {
+    this.setState({
+      todos: filterTodos(this.state.todos, Filter.ACTIVE)
```
Same here

trevorwright commented on src/components/App.js:
```diff
@@ -31,37 +31,95 @@ function filterTodos(todos, filter) {
+  }
+
+  render() {
+    const {todos, selectedFilter} = this.state;
```
nit: space between your curlies and the items within... `{ todos, selectedFilter }`

trevorwright commented on src/components/App.js:
```diff
@@ -31,37 +31,95 @@ function filterTodos(todos, filter) {
+        </AppTitleStyled>
+
+        <ContentStyled>
+          <Header onEnterNewTodo={this.onEnterNewTodo} onUpdateAllCompletedTo={this.onUpdateAllCompletedTo} markAllCompleteChecked={todos.length > 0 && todos.every(todo => todo.completed)}/>
```
These are some really long lines and should be cleaned up. Our coding standard is 100 line width. You can turn on the ruler in VSCode to see what that looks like

trevorwright replied:
In JavaScript land you can drop the `> 0` here 👍. You may also want to factor out this logic for determining if mark all complete checked into its own function

trevorwright commented on src/components/Footer.js:
```diff
@@ -32,15 +32,15 @@ const ClearCompletedStyled = styled.a`
 `;
 
-const Footer = ({ todosRemainingCount, selectedFilter }) => (
+const Footer = ({ todosRemainingCount, selectedFilter, onUpdateSelectedFilter, onClearCompleted }) => (
```
You've added props, but prop-types have not been updated. There are prop types for expecting a func.

trevorwright commented on src/components/Header.js:
```diff
@@ -47,11 +47,36 @@ const NewTodoFieldStyled = styled.input`
+    }
+
+    onInputChange = (event) => {
+        this.setState({newToDoTitle: event.target.value});
```
nit: spacing with curlies as mention above 👻

trevorwright commented on src/components/Header.js:
```diff
@@ -47,11 +47,36 @@ const NewTodoFieldStyled = styled.input`
+            this.props.onEnterNewTodo(this.state.newToDoTitle); 
+
+            // Clear input
+            this.setState({ newToDoTitle: ''});
```
nit: curlies 🏂

trevorwright commented on src/components/Header.js:
```diff
@@ -47,11 +47,36 @@ const NewTodoFieldStyled = styled.input`
+        this.setState({newToDoTitle: event.target.value});
+    };
+
+    onKeyPress = (event) => {
```
There is an argument to be made here that handling the clearing of the input is not the responsibility of the `onKeyPress` function, so you'd say if 13 then call the function which handles doing that stuff.

I don't really care for this app, but just worth thinking about as you write 👍

trevorwright commented on src/components/Header.js:
```diff
@@ -47,11 +47,36 @@ const NewTodoFieldStyled = styled.input`
-);
+const Header = class extends React.Component {
+    constructor(props) {
+        super(props);
```
Your props should be described with prop-types

trevorwright commented on src/components/ToDoItem.js:
```diff
@@ -53,27 +53,47 @@ const ToDoEditStyled = styled.input`
+    constructor(props) {
+        super(props);
+
+        this.state = {
```
Same comment as Joe was making above

trevorwright commented on src/components/ToDoItem.js:
```diff
@@ -53,27 +53,47 @@ const ToDoEditStyled = styled.input`
+
+    onEditKeyPress = (event) => {
+        if (event.which === 13) {
+            this.props.onUpdateToDo(this.props.todo, { ...this.props.todo, title: this.state.editText, editing: false });
```
I would argue that you determine the second param on its own line and pass it into the function for readability.

trevorwright commented on src/components/ToDoItem.js:
```diff
@@ -53,27 +53,47 @@ const ToDoEditStyled = styled.input`
-    )
+    render() {
+        const {
+            props: { todo, onDeleteToDo, onToggleToDoComplete, onEditToDo, onStartEditTodo, onUpdateToDo },
```
These should all be added to prop-types

trevorwright commented on src/components/ToDoItem.js:
```diff
@@ -53,27 +53,47 @@ const ToDoEditStyled = styled.input`
+        let todoItem;
+
+        if (todo.editing) {
+            todoItem = <ToDoEditStyled type="text" autoFocus value={this.state.editText} onChange={(event) => this.setState({ editText: event.target.value })} onKeyPress={this.onEditKeyPress}/>
```
Long line

trevorwright commented on src/components/ToDoItem.js:
```diff
@@ -53,27 +53,47 @@ const ToDoEditStyled = styled.input`
+        } else {
+            todoItem =
+                <Fragment>
+                    <CompleteToggleStyled type="checkbox" checked={todo.completed} onChange={() => onUpdateToDo(todo, { ...todo, completed: !todo.completed})} />
```
There should probably be a `toggleTodoCompleted` function which handles this logic.

trevorwright commented on src/components/ToDoItem.js:
```diff
@@ -53,27 +53,47 @@ const ToDoEditStyled = styled.input`
+            todoItem =
+                <Fragment>
+                    <CompleteToggleStyled type="checkbox" checked={todo.completed} onChange={() => onUpdateToDo(todo, { ...todo, completed: !todo.completed})} />
+                    <ToDoItemLabelStyled onDoubleClick={() => {onUpdateToDo(todo, {...todo, editing: true})}}>
```
Same as above. You want to try and keep your JSX with most business logic outside for readability.

trevorwright reviewed (changes requested):
This is looking good. Comments are generally minor and a lot of them are just getting used to our styles of coding and the ins/outs of React.

🥇 🎉

musharraf-pg committed:
Converting setState to use updater function and other changes

musharraf-pg commented:
Have hopefully addressed all comments besides whitespace stuff. Once this PR is approved, I will create a new PR that fixes all whitespace and style stuff which Prettier is now screaming about.

musharraf-pg committed:
Moved state initialization from constructor to class property

trevorwright commented on src/components/App.js:
```diff
@@ -31,37 +31,114 @@ function filterTodos(todos, filter) {
+  onEnterNewTodo = (title) => {
+    this.setState((prevState) => {
+        return {
+            todos: prevState.todos.concat([{ title, completed: false, editing: false, id: uniqueId() }])
```
You can also use magical ES6 syntax to make this super awesome.

```
this.setState(({ todos }) => ({ todos: [...todos, { <new object stuff }]);
```

trevorwright commented on src/components/App.js:
```diff
@@ -31,37 +31,114 @@ function filterTodos(todos, filter) {
+  };
+
+  onDeleteToDo = (todo) => {
+    this.setState((prevState) => {
```
Same here, you can deconstruct specific values so instead of prevState just get the prop you need. It makes code nice and clean.

`this.setState(({ todos })`

trevorwright commented on src/components/App.js:
```diff
@@ -31,37 +31,114 @@ function filterTodos(todos, filter) {
+    });
+  };
+
+  // onToggleToDoComplete = (todo) => {
```
Don't commit commented code

trevorwright commented on src/components/App.js:
```diff
@@ -31,37 +31,114 @@ function filterTodos(todos, filter) {
+  };
+
+  onUpdateSelectedFilter = (selectedFilter) => {
+      this.setState((prevState) => {
```
You don't need to use functional set state here, because you are not dependant on the previous state

trevorwright commented on src/components/App.js:
```diff
@@ -31,37 +31,114 @@ function filterTodos(todos, filter) {
+  onUpdateAllCompletedTo = (newCompletedStatus) => {
+      this.setState((prevState) => {
+          return {
+            todos: prevState.todos.map(todo => ({...todo, completed: newCompletedStatus}))
```
don't forget that white space with curlies ;)

musharraf-pg replied:
yup will fix all whitespace and curlies issues with running prettier on each file in next PR. just didn't want to do it here to avoid muddling this PR with formatting changes which will affect nearly every other line

trevorwright commented on src/components/App.js:
```diff
@@ -31,37 +31,114 @@ function filterTodos(todos, filter) {
   );
+
+  render() {
+    const {todos, selectedFilter} = this.state;
```
white space 👻

trevorwright commented on src/components/ToDoList.js:
```diff
@@ -11,16 +11,18 @@ const ToDoListStyled = styled.ul`
 `;
 
-const ToDoList = ({ todos }) => (
+const ToDoList = ( { todos, ...rest } ) => (
```
No whitespace outside curlies 👍

trevorwright reviewed (approved):
Just some small things but this is lookin' good!

musharraf-pg committed:
Moved business logic outside of JSX into component methods

musharraf-pg committed:
Destructuring in various places
